sdk/python/kfp/components/_python_op.py

```diff
@@ -45,6 +45,47 @@ def _python_function_name_to_component_name(name):
     return re.sub(' +', ' ', name.replace('_', ' ')).strip(' ').capitalize()
 
 
+def _capture_function_code_using_source_copy(func) -> str:
+    import inspect
+
+    #Source code can include decorators line @python_op. Remove them
+    (func_code_lines, _) = inspect.getsourcelines(func) 
+    while func_code_lines[0].lstrip().startswith('@'): #decorator
+        del func_code_lines[0]
+
+    #Function might be defined in some indented scope (e.g. in another function).
+    #We need to handle this and properly dedent the function source code
+    first_line = func_code_lines[0]
+    indent = len(first_line) - len(first_line.lstrip())
+    func_code_lines = [line[indent:] for line in func_code_lines]
+
+    #TODO: Add support for copying the NamedTuple subclass declaration code
+    #Adding NamedTuple import if needed
+    if hasattr(inspect.signature(func).return_annotation, '_fields'): #NamedTuple
+        func_code_lines.insert(0, '\n')
+        func_code_lines.insert(0, 'from typing import NamedTuple\n')
+
+    return ''.join(func_code_lines) #Lines retain their \n endings
+
+
+def _capture_function_code_using_cloudpickle(func) -> str:
+    import sys
+    import cloudpickle
```

Comment:
This looks hacky to me. It uses cloudpickle to serialize the function along with its dependencies and uses pickle to deserialize it. What if the formats output from the cloudpickle are no longer consistent with the pickle input in the future? Depending on two extra packages to be consistent with each other does not seem to be sustainable. 

However, I'm all up for resolving the lightweight component limitations.

Reply:
>It uses cloudpickle to serialize the function along with its dependencies and uses pickle to deserialize it. 

That's how cloudpickle works. Cloudpickle is just an extension of pickle. It's even inherited from it. It just overrides several methods to add serialization of several additional types and cases. Cloudpickle does **not** have any deserializers. It serializes to pickle format using pickle itself.